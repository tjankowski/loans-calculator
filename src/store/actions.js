import quote from "../api";
import { FormFieldsValueRange, FormFields } from "../store/constants";
import debounce from "lodash.debounce";

export const actionTypes = {
  QUOTE_FETCH: "quote.fetch",
  QUOTE_SUCCESS: "quote.success",
  QUOTE_ERROR: "quote.error",
  FORM_INPUT: "form.input",
};

export const useActions = (state, dispatch) => ({
  quote: (amount, duration) => {
    return fetchQuote(dispatch, { amount, duration });
  },
  onFormChange: (field, input) => {
    const value = validateFormInput(field, input);
    dispatch(action(actionTypes.FORM_INPUT, { field, value }));
    if (!value.isError) {
      return fetchQuote(dispatch, {
        amount: state.form[FormFields.amount].value,
        duration: state.form[FormFields.duration].value,
        [field]: value.value,
      });
    }
    return null;
  },
});

const fetchQuote = debounce((dispatch, { amount, duration }) => {
  dispatch(action(actionTypes.QUOTE_FETCH, { amount, duration }));
  return quote(amount, duration)
    .then((data) => {
      dispatch(action(actionTypes.QUOTE_SUCCESS, data));
    })
    .catch((error) => {
      dispatch(action(actionTypes.QUOTE_ERROR, error));
    });
}, 200);

function validateFormInput(field, value) {
  const [min, max] = FormFieldsValueRange[field];
  return {
    isError: value < min || value > max,
    value: Number(value > max ? max : value),
  };
}

export function action(type, payload) {
  return {
    type,
    payload,
  };
}
