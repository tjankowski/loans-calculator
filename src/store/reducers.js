import { actionTypes } from "./actions";
import { FormFields, Currency } from "./constants";

export const initialState = {
  quote: {
    isFetching: false,
    isError: false,
    data: null,
  },
  currency: Currency.GBP,
  form: {
    [FormFields.amount]: {
      value: 2000,
      isError: false,
    },
    [FormFields.duration]: {
      value: 24,
      isError: false,
    },
  },
};

export function reducer(state = initialState, action = { type: null }) {
  switch (action?.type) {
    case actionTypes.FORM_INPUT:
      return {
        ...state,
        quote: {
          ...state.quote,
          data: action.payload.value.isError ? null : state.quote.data,
        },
        form: {
          ...state.form,
          [action.payload.field]: action.payload.value,
        },
      };
    case actionTypes.QUOTE_ERROR:
      return {
        ...state,
        quote: {
          ...state.quote,
          isError: true,
          isFetching: false,
          data: action.payload,
        },
      };

    case actionTypes.QUOTE_SUCCESS:
      return {
        ...state,
        quote: {
          ...state.quote,
          isError: false,
          isFetching: false,
          data: action.payload,
        },
      };
    case actionTypes.QUOTE_FETCH:
      return {
        ...state,
        quote: {
          ...state.quote,
          isError: false,
          isFetching: true,
        },
      };
    default:
      return state;
  }
}
