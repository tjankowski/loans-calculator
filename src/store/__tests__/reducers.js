import { reducer, initialState } from "../reducers";
import { actionTypes, action } from "../actions";

describe("Initial state:", () => {
  it("should have valid inital state", () => {
    expect(initialState).toMatchSnapshot();
  });

  it("should return same state for without action", () => {
    expect(reducer(initialState)).toMatchSnapshot();
  });

  it("should return initial state state without arguments", () => {
    expect(reducer()).toMatchSnapshot();
  });
});

describe("For actions:", () => {
  it("should return same state for unknown action", () => {
    expect(reducer(initialState, action(Date.now()))).toMatchSnapshot();
  });

  it("should set fetching state", () => {
    expect(
      reducer(initialState, action(actionTypes.QUOTE_FETCH))
    ).toMatchSnapshot();
  });

  it("should set fetching error state", () => {
    expect(
      reducer(
        initialState,
        action(actionTypes.QUOTE_ERROR, { error: "Error message" })
      )
    ).toMatchSnapshot();
  });

  it("should set fetching success state", () => {
    expect(
      reducer(
        initialState,
        action(actionTypes.QUOTE_SUCCESS, { data: [1, 2, 3] })
      )
    ).toMatchSnapshot();
  });
});
