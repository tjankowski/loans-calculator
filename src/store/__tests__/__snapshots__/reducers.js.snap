// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`For actions: should return same state for unknown action 1`] = `
Object {
  "currency": Object {
    "key": "GBP",
    "label": "£",
  },
  "form": Object {
    "amount": Object {
      "isError": false,
      "value": 2000,
    },
    "duration": Object {
      "isError": false,
      "value": 24,
    },
  },
  "quote": Object {
    "data": null,
    "isError": false,
    "isFetching": false,
  },
}
`;

exports[`For actions: should set fetching error state 1`] = `
Object {
  "currency": Object {
    "key": "GBP",
    "label": "£",
  },
  "form": Object {
    "amount": Object {
      "isError": false,
      "value": 2000,
    },
    "duration": Object {
      "isError": false,
      "value": 24,
    },
  },
  "quote": Object {
    "data": Object {
      "error": "Error message",
    },
    "isError": true,
    "isFetching": false,
  },
}
`;

exports[`For actions: should set fetching state 1`] = `
Object {
  "currency": Object {
    "key": "GBP",
    "label": "£",
  },
  "form": Object {
    "amount": Object {
      "isError": false,
      "value": 2000,
    },
    "duration": Object {
      "isError": false,
      "value": 24,
    },
  },
  "quote": Object {
    "data": null,
    "isError": false,
    "isFetching": true,
  },
}
`;

exports[`For actions: should set fetching success state 1`] = `
Object {
  "currency": Object {
    "key": "GBP",
    "label": "£",
  },
  "form": Object {
    "amount": Object {
      "isError": false,
      "value": 2000,
    },
    "duration": Object {
      "isError": false,
      "value": 24,
    },
  },
  "quote": Object {
    "data": Object {
      "data": Array [
        1,
        2,
        3,
      ],
    },
    "isError": false,
    "isFetching": false,
  },
}
`;

exports[`Initial state: should have valid inital state 1`] = `
Object {
  "currency": Object {
    "key": "GBP",
    "label": "£",
  },
  "form": Object {
    "amount": Object {
      "isError": false,
      "value": 2000,
    },
    "duration": Object {
      "isError": false,
      "value": 24,
    },
  },
  "quote": Object {
    "data": null,
    "isError": false,
    "isFetching": false,
  },
}
`;

exports[`Initial state: should return initial state state without arguments 1`] = `
Object {
  "currency": Object {
    "key": "GBP",
    "label": "£",
  },
  "form": Object {
    "amount": Object {
      "isError": false,
      "value": 2000,
    },
    "duration": Object {
      "isError": false,
      "value": 24,
    },
  },
  "quote": Object {
    "data": null,
    "isError": false,
    "isFetching": false,
  },
}
`;

exports[`Initial state: should return same state for without action 1`] = `
Object {
  "currency": Object {
    "key": "GBP",
    "label": "£",
  },
  "form": Object {
    "amount": Object {
      "isError": false,
      "value": 2000,
    },
    "duration": Object {
      "isError": false,
      "value": 24,
    },
  },
  "quote": Object {
    "data": null,
    "isError": false,
    "isFetching": false,
  },
}
`;
